text/0000-let-else.md

```diff
@@ -0,0 +1,473 @@
+
+The `else` block must diverge. This could be a keyword which diverges (returns `!`), or a panic.
+This likely necessitates a new subtype of `BlockExpression`, something like `BlockExpressionDiverging`.
+Allowed keywords:
+- `return`
+- `break`
+- `continue`
+
+If the pattern does not match, the expression is not consumed, and so any existing variables from the surrounding scope are
+accessible as they would normally be.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## The diverging block
+
+"Must diverge" is an unusual requirement, which doesn't exist elsewhere in the language as of the time of writing, 
+and might be difficult to explain or lead to confusing errors for programmers new to this feature.
+
+This also necessitates a new block expression subtype, something like `BlockExpressionDiverging`.
+
+## `let PATTERN = if {} else {} else {};`
+
+One unfortunate combination of this feature with regular if-else expressions is the possibility of `let PATTERN = if { a } else { b } else { c };`.
+This is likely to be unclear if anyone writes it, but does not pose a syntactical issue, as `let PATTERN = if y { a } else { b };` should always be
+interpreted as `let Enum(x) = (if y { a } else { b });` (still a compile error as there no diverging block: `error[E0005]: refutable pattern in local binding: ...`)
+because the compiler won't interpret it as `let PATTERN = (if y { a }) else { b };` since `()` is not an enum.
+
+This can be overcome by making a raw if-else in the expression position a compile error and instead requiring that parentheses are inserted to disambiguate:
+`let PATTERN = (if { a } else { b }) else { c };`.
+
+Rust already provides us with such a restriction, and so the expression can be restricted to be a [`ExpressionWithoutBlock`][expressions].
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+let-else attempts to be as consistent as possible to similar existing syntax.
+
+Fundamentally it is treated as a `let` statement, necessitating an assignment and the trailing semicolon.
+
+Pattern matching works identically to if-let, no new "negation" pattern matching rules are introduced.
+
+Operator precedence with `&&` in made to be like if-let, requiring that a case which is an error prior to this RFC be changed to be a slightly different error.
+This is for a possible extension for let-else similar to the (yet unimplemented) if-else-chains feature, as mentioned in [future-possibilities][] with more detail.
+Specifically, while the following example is an error today, by the default `&&` operator rules it would cause problems with if-let-chains like `&&` chaining:
+
+```rust
+let a = false;
+let b = false;
+
+// The RFC proposes boolean matches like this be either:
+// - Made into a compile error, or
+// - Made to be parsed internally like if-let-chains: `(let true = a) && b else { ... };`
```

Comment:
What would this parsing mean, exactly? `b else` doesn't seem like it parses.

Reply:
The syntax is intended to be `((let true = a) && b) else { diverge };` (where the expression on the left is required to be an `&&`-separated list of `let PAT = EXPR` and `EXPR` terms) and if `a` matches `true` and `b` is true then it goes to the next statement (and binds any patterns in the lets) otherwise it runs `diverge`. I am also dubious about the use of parentheses here, and it's quite possible that the outer set of parentheses will not be permitted.

Reply:
I suppose it would really be more like:
```
<known that it is a let statement> ((let true = a) && b) else { ... }
```

I don't know how to express this parsing in natural rust without some kind of extra internal compiler directive that it is a lest statement.

The rules would otherwise be as per the [if-let-chains](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md) rfc.

Reply:
I think it'd be simpler to say that `let true = a && b else { ... }` should just not parse for now, to allow for forward-compatibility. I don't think at this time that we need a mechanism to combine let-else with non-pattern-match boolean conditions; we just need to acknowledge that there would be a conflict there.

I *can* imagine someone wanting to combine `if let` with `let else`, but that seems *really* complicated. For now, let's just say that they don't combine at all.

Reply:
Accepted the decision to make that case an error and also added another future possibilities note for that last _unfortunate_ case.